### PR TITLE
Support atoms with multiple types

### PR DESCRIPTION
**Pull request by jennybuckley**
Get rid of schema.DeduceType and instead allow atoms with multiple types defined and use the type of data to eliminate incompatible types.

Also changes Untyped Deduced to this:
```
types:
- name: __untyped_deduced_
  scalar: untyped
  list:
    elementType:
      namedType: __untyped_atomic_
    elementRelationship: atomic
  map:
    elementType:
      namedType: __untyped_deduced_
    elementRelationship: separable
```

This is necessary if we want to support things like JSONPropsOrArray in the future

### CODE REVIEW

File: typed/merge.go [hunk 1]
```diff
@@ -374,8 +392,8 @@ func (w *mergingWalker) visitMapItems(t schema.Map, lhs, rhs *value.Map) (errs V
 
 func (w *mergingWalker) doMap(t schema.Map) (errs ValidationErrors) {
 	var lhs, rhs *value.Map
-	errs = append(errs, w.derefMapOrStruct("lhs: ", "map", w.lhs, &lhs)...)
-	errs = append(errs, w.derefMapOrStruct("rhs: ", "map", w.rhs, &rhs)...)
+	w.derefMapOrStruct("lhs: ", "map", w.lhs, &lhs)
+	w.derefMapOrStruct("rhs: ", "map", w.rhs, &rhs)
```
Comment (jennybuckley) [inline typed/merge.go]: We could actually remove these errors from all the cases, but I wanted to keep the diff smaller

Comment (apelisse) [reply typed/merge.go]: If we don't think that they should ever happen, can we panic?

Comment (jennybuckley) [reply typed/merge.go]: the error case can happen, but what we want it to do in that case is treat the value as nil

File: typed/merge.go [hunk 2]
```diff
@@ -61,12 +63,26 @@ var (
...
-	errs := resolveSchema(w.schema, w.typeRef, w)
+	a, ok := w.schema.Resolve(w.typeRef)
+	if !ok {
+		return w.errorf("schema error: no type found matching: %v", *w.typeRef.NamedType)
+	}
+
+	alhs := deduceAtom(a, w.lhs)
+	arhs := deduceAtom(a, w.rhs)
+	if reflect.DeepEqual(alhs, arhs) {
+		errs =append(errs, handleAtom(arhs, w.typeRef, w)...)
```
Comment (apelisse) [inline typed/merge.go]: very important: missing space :-)

**Comment by apelisse** [comment]
Do you want to squash and I lgtm or do you want to move the type to schemaconv?

**Comment by apelisse** [comment]
I think we agree that it was better before that last commit?

**Comment by apelisse** [comment]
We need to add a little bit more testing for this feature, let's follow-up after.
Also we need to fix the gofmt thing, I'll probably force merge.

**Pull request merged**